### ISSUE TITLE
print() and print_when() could be more ergonomic to use

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#8176]
These are great and essential tools for debugging, but they could be a lot better:
- They are painful to conditionally insert into the code because the syntax is very intrusive
- Turning them on and off requires recompiles

Thought: what if we added a scheduling directive for `print_when()` that would inject the print when the func is evaluated? (No need for print() since that's just print_when(true))? And added some optional controls via env var (eg print_when_env("ENV_VAR_NAME")?)

**Comment by abadams** [comment]
So an unconditional print is just Func::trace_stores. We could add an optional boolean Expr arg to it?

**Comment by steven-johnson** [comment]
> So an unconditional print is just Func::trace_stores. We could add an optional boolean Expr arg to it?

Yes and know; when I use these for debugging I usually also want to print other data too
